ocaml/typing/typeopt.ml

```diff
@@ -480,31 +524,44 @@ and value_kind_record env ~loc ~visited ~depth ~num_nodes_visited
     end
 
 let value_kind env loc ty =
-  let (_num_nodes_visited, value_kind) =
-    value_kind env ~loc ~visited:Numbers.Int.Set.empty ~depth:0
-      ~num_nodes_visited:0 ty
-  in
-  value_kind
+  try
+    let (_num_nodes_visited, value_kind) =
+      value_kind env ~loc ~visited:Numbers.Int.Set.empty ~depth:0
+        ~num_nodes_visited:0 ty
+    in
+    value_kind
+  with
+  | Missing_cmi_fallback -> raise (Error (loc, Non_value_layout (ty, None)))
 
-(* CR layouts v2: We are planning to put a sanity check here that you don't get
-   passed float# as the sort unless layouts_alpha is on.  This violates our
-   previous rule that extensions only control syntax, but seems like a nice
-   implementation of a sanity check, since the availability of the Float_u
-   module will result in ways to get at unboxed floats that aren't just writing
-   #float in your program.
+(* CR layouts v2: We are planning to adjust the sanity check here to allow
+   #float only if layouts_alpha is on.  This violates our previous rule that
+   extensions only control syntax, but seems like a nice implementation of a
+   sanity check, since the availability of the Float_u module will result in
+   ways to get at unboxed floats that aren't just writing #float in your
+   program.
 
    We also do the void sanity check here.  We do it in value_kind as well,
    because the sort argument here is sometimes not computed from the type or
    typed tree, but just one of the defaults from layouts.ml.
 *)
 let layout env loc sort ty =
   match Layouts.Sort.get_default_value sort with
-  | Void -> raise (Error (loc, Non_value_sort ty))
+  | Void -> raise (Error (loc, Non_value_sort (Sort.void,ty)))
+  | Float64 when Language_extension.(is_at_least Layouts Alpha) ->
+    Lambda.Punboxed_float
+  | Float64 -> raise (Error (loc, Non_value_sort (Sort.float64,ty)))
   | Value -> Lambda.Pvalue (value_kind env loc ty)
 
 let layout_of_sort loc sort =
   match Layouts.Sort.get_default_value sort with
-  | Void -> raise (Error (loc, Non_value_sort_unknown_ty))
+  | Void -> raise (Error (loc, Non_value_sort_unknown_ty Sort.void))
+  | Float64 -> Lambda.Punboxed_float
+  | Value -> Lambda.Pvalue Pgenval
+
+let layout_of_const_sort (s : Layouts.Sort.const) =
+  match s with
+  | Void -> assert false
+  | Float64 -> Lambda.Punboxed_float
```

Comment:
This one probably doesn't need it as badly, but may as well check the language extension here too.